### PR TITLE
Add AppendFormat extension methods on IHtmlContent

### PR DESCRIPTION
**Pull request by rynowak**
/cc @NTaylorMullen - this implements https://github.com/aspnet/Razor/issues/526

### CODE REVIEW

File: src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs [hunk 1]
```diff
@@ -14,6 +16,52 @@ namespace Microsoft.AspNet.Html.Abstractions
...
+        /// <param name="args">
+        /// The object array to format. Each element in the array will be formatted and then HTML encoded.
+        /// </param>
+        /// <returns>A reference to this instance after the append operation has completed.</returns>
+        public static IHtmlContentBuilder AppendFormat(
+            this IHtmlContentBuilder builder,
+            string format,
+            params object[] args)
+        {
+            builder.Append(new HtmlFormatString(format, args));
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: argument `null` check for builder.

File: src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs [hunk 2]
```diff
@@ -14,6 +16,52 @@ namespace Microsoft.AspNet.Html.Abstractions
...
+        /// The object array to format. Each element in the array will be formatted and then HTML encoded.
+        /// </param>
+        /// <returns>A reference to this instance after the append operation has completed.</returns>
+        public static IHtmlContentBuilder AppendFormat(
+            this IHtmlContentBuilder builder,
+            IFormatProvider formatProvider,
+            string format,
+            params object[] args)
+        {
+            builder.Append(new HtmlFormatString(formatProvider, format, args));
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: ditto

File: src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs [hunk 3]
```diff
@@ -132,5 +180,145 @@ private string DebuggerToString()
...
+                }
+
+                if (encoder == null)
+                {
+                    throw new ArgumentNullException(nameof(encoder));
+                }
+
+                var formatProvider = new EncodingFormatProvider(_formatProvider, encoder);
+                var builder = new StringBuilder();
+                builder.AppendFormat(formatProvider, _format, _args);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: `builder.AppendFormat` can't handle a `null` value and we don't `null` check it in the ctor.

File: src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs [hunk 4]
```diff
@@ -132,5 +180,145 @@ private string DebuggerToString()
...
+
+                for (var i = 0; i < builder.Length; i++)
+                {
+                    writer.Write(builder[i]);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: :sob: really wish `StringBuilder` could be treated as a `char[]`

File: src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs [hunk 5]
```diff
@@ -132,5 +180,145 @@ private string DebuggerToString()
...
+        // This class implements Html encoding via an ICustomFormatter. Passing an instance of this
+        // class into a string.Format method or anything similar will evaluate arguments implementing
+        // IHtmlContent without HTML encoding them, and will give other arguments the standard
+        // composite format string treatment, and then HTML encode the result.
+        //
+        // Plenty of examples of ICustomFormatter and the interactions with string.Format here:
+        // https://msdn.microsoft.com/en-us/library/system.string.format(v=vs.110).aspx#Format6_Example
+        private class EncodingFormatProvider : IFormatProvider, ICustomFormatter
+        {
+            private IHtmlEncoder _encoder;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: readonly here and in next line.

File: src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs [hunk 6]
```diff
@@ -132,5 +180,145 @@ private string DebuggerToString()
...
+
+            public EncodingFormatProvider(IFormatProvider formatProvider, IHtmlEncoder encoder)
+            {
+                _formatProvider = formatProvider;
+                _encoder = encoder;
+            }
+
+            public string Format(string format, object arg, IFormatProvider formatProvider)
+            {
+                // This is the case we need to special case, and avoid encoding.
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: How does this bit avoid encoding? We're just calling through to `.WriteTo` on an `IHtmlContent` which may or may not encode.

File: src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs [hunk 7]
```diff
@@ -132,5 +180,145 @@ private string DebuggerToString()
...
+                    if (result != null)
+                    {
+                        return _encoder.HtmlEncode(result);
+                    }
+                }
+
+                // Next check if 'arg' is an IFormattable (DateTime is an example).
+                //
+                // An IFormattable will likely call back into the IFormatterProvider and ask for more information
+                // about how to format itself. This is the typical case when IFormatterProvider is an CultureInfo.
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: `a CultureInfo.`

File: src/Microsoft.AspNet.Html.Abstractions/project.json [hunk 8]
```diff
@@ -16,7 +16,7 @@
         "dnx451": { },
         "dnxcore50": {
             "dependencies": {
-                "System.Resources.ResourceManager": "4.0.1-beta-"
+                "System.Resources.ResourceManager": "4.0.1-beta-*"
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Html.Abstractions/project.json]: :+1:

**Comment by NTaylorMullen** [comment]
:watc

File: src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: It avoids double encoding. The `IHtmlContent` does the right thing

Comment (rynowak) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: I'll make the comment better.

File: src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: :+1:

**Comment by rynowak** [comment]
@NTaylorMullen updated

**Comment by NTaylorMullen** [comment]
:shipit:

File: src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs (hunk 3 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: Isn't `string.Format` equivalent?

Comment (rynowak) [inline src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs]: Yeah I suppose we're not buying much by doing it this way. I'll simplify it.

**Pull request closed without merging**